src/frequenz/sdk/microgrid/_data_pipeline.py

```diff
@@ -119,29 +118,22 @@ def __init__(
         self._grid: Grid | None = None
         self._ev_charger_pools: dict[frozenset[int], EVChargerPool] = {}
         self._battery_pools: dict[frozenset[int], BatteryPoolReferenceStore] = {}
-        self._frequency_pool: dict[int, GridFrequency] = {}
+        self._frequency_instance: GridFrequency
 
-    def frequency(self, component: Component | None = None) -> GridFrequency:
+    def frequency(self) -> GridFrequency:
         """Fetch the grid frequency for the microgrid.
 
-        Args:
-            component: The component to use when fetching the grid frequency.  If None,
-                the component will be fetched from the registry.
-
         Returns:
-            A GridFrequency instance.
+            The GridFrequency instance.
         """
-        if component is None:
-            component = GridFrequency.find_frequency_component()
+        if not hasattr(self, "_frequency_instance"):
```

Comment:
Why the use of this dynamic idiom? Why not make `_frequency_instance: GridFrequency | None`?

Reply:
You mean completely forbid `{has,get,set}attr` without a `# pylint: disable=/noqa`? :thinking:

Reply:
yup,  because those require putting code in strings, which bypasses our checks.  So they should be manually excluded and documented where ever they are used.

Reply:
I suppose I wanted to avoid extending the type to `None` thus requiring extra checks for whether it's set or not..
In this context it doesn't really make a difference I suppose.